File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <iostream>
#include <stack>

using namespace std;

int open = 0;
int close = 0;

bool solution(string s)
{
    bool answer = true; 
    stack<char> st;
    
    if(s.size() % 2 != 0) return false;
    
    for(int i=0; i<s.size(); i++){
        if(st.empty()){
            if(s[i] == ')') return false;
            else st.push(s[i]);
        }else {
            if(s[i] == ')'){
                if(!st.empty() && st.top() == '(') st.pop();
            }else {
                st.push(s[i]);
            }
        }
    }
    if(!st.empty()) answer = false;

    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.Stack;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        Stack<Character> st = new Stack<>();
        
        if (s.length() % 2 != 0)
            return false;
        
        for (int i = 0; i < s.length(); i++) {
            if (st.isEmpty()) {
                if (s.charAt(i) == ')')
                    return false;
                else
                    st.push(s.charAt(i));
            } else {
                if (s.charAt(i) == ')') {
                    if (!st.isEmpty() && st.peek() == '(')
                        st.pop();
                } else {
                    st.push(s.charAt(i));
                }
            }
        }
        if (!st.isEmpty())
            answer = false;

        return answer;
    }
}